### ISSUE TITLE
Increase functionality and efficiency to Spritebatch via alterations to the Spritefont class

### ISSUE THREAD

**Issue by willmotil** [MonoGame/MonoGame#4812]
This maybe a bit early for this post but in consideration of the proposed addition of emoji characters. In #4805 i suppose i should post this as Spritefont is up for changes

Ill begin by stating i have reduced the spritefont.DrawInto methods matrix calculations. 
As a result of trying to allow for a culling boundary rectangle to be added to a custom drawString method. I have come to realize, the Spritefont class is a bit poorly structured in relation to Spritebatch.

https://drive.google.com/open?id=0B1zD887frY04bkJkU3hEMmZoT1k

Internal Measurestring, Drawinto, Charsource, are tightly coupled and both the Charsource struct in Spritefont and Drawinto are tightly coupled to Spritebatch. 
Basically Spritefont holds some Spritebatches responsibility were it should not.

If we could bypass DrawInto with our own Drawstring then pass the results to DrawInternal.. 
From this we could then create our own custom methods that manipulate text in unique ways.

<img src="http://community.monogame.net/uploads/default/original/2X/b/b746a8b40b72d3c734e7191dd321ccfaeac81c08.png">

<img src="http://community.monogame.net/uploads/default/optimized/2X/8/806c970c107ae2f004b08b979e9853e3b6c329d6_1_486x499.png">

Even if the above doesn't sound good drawinto could use a tune up.

I propose the following be considered, 
The charsource struct.
The internal measurestring .
The method Drawinto 
All be moved to the Spritebatch class. 

Further that DrawInto is simply a internal Drawstring and it be possibly renamed to reflect that. 

Also considering the possibility that the struct Charsource maybe redundant as well as it is not preventing garbage from being produced and collected.

**Comment by willmotil** [comment]
Here is a draft DrawInto method Unrolled some of it is not quite the same for testing but essentially its just Spritefont.DrawInto with the matrix's unrolled it comes out to about 8 mults and 10 addition subtractions maximum less if your not flipping or rotating it still needs to be touched up but its a lot simpler to see exactly what is going on with it now.

`        
        // working test drft
        public class SpritefontOrSpriteBatch
        {
            private Dictionary<char, SpriteFont.Glyph> _glyphs;// = Engien.GetFont.GetGlyphs();
            private SpriteFont.Glyph defaultGlyph;
            private SpriteFont tsf;
            private char defaultfontchar = ' ';

            public void SetSpriteFont(SpriteFont s)
            {
                tsf = s;
                _glyphs = tsf.GetGlyphs();
                defaultGlyph = new SpriteFont.Glyph();
                if (tsf.DefaultCharacter.HasValue)
                {
                    defaultfontchar = (char)(tsf.DefaultCharacter.Value);
                    defaultGlyph = _glyphs[defaultfontchar];
                }
            }

            /// <summary>
            /// This method does Not call DrawString to draw text it calls Draw.
            /// This method directly simulates SpriteFontDrawInto without using matrices.
            /// this example basically unrolls Spritefont.Drawinto
            /// </summary>
            public void DrawText(SpriteBatch spriteBatch, SpriteFont tsf, StringBuilder text, Vector2 position, Color color,
                    float rotation, Vector2 origin, Vector2 scale, SpriteEffects effect, float depth, Rectangle textboundryarea)
            {
                var LineSpacing = (float)tsf.LineSpacing;
                var Spacing = tsf.Spacing;
                bool flip = false;
                bool flippedVert = false;
                bool flippedHorz = false;
                Vector2 msize = Vector2.Zero;
                Vector2 offset = Vector2.Zero;
                Vector2 q = new Vector2((float)(Math.Sin(rotation)), (float)(Math.Cos(rotation)));
                Rectangle dest = new Rectangle();
                var currentGlyph = defaultGlyph;
                var firstGlyphOfLine = true;

                // set flag if we can skip all the spriteeffect calculations
                if (effect != SpriteEffects.None)
                {
                    flip = true;
                    // measure string needs fixed next its terrible slow and its not the same
                    msize = tsf.MeasureString(text);
                    flippedHorz = (effect & SpriteEffects.FlipHorizontally) == SpriteEffects.FlipHorizontally;
                    flippedVert = (effect & SpriteEffects.FlipVertically) == SpriteEffects.FlipVertically;
                }

                // begin looping thru text
                for (var i = 0; i < text.Length; i++)
                {
                    var c = text[i];
                    //int isc = char.ConvertToUtf32(c, c);

                    if (c == '\r')
                        continue;
                    if (c == '\n')
                    {
                        offset.X = 0;
                        offset.Y += LineSpacing;
                        firstGlyphOfLine = true;
                        continue;
                    }

                    // The original version is prolly better on this part i dunno.
                    // I just chucked this in to save time, maybe the default should be a red square.
                    // using the first glyph size and drop in a dot texture with bounds 0,0,1,1
                    // it should be precomputed in the spritefonts constructor once either way
                    if (_glyphs.ContainsKey(c))
                        currentGlyph = _glyphs[c];
                    else
                        if (!tsf.DefaultCharacter.HasValue)
                            throw new ArgumentException("Text Contains a Unresolvable Character");
                        else
                            currentGlyph = defaultGlyph;

                    // Solves the problem- the first character on a line with a negative left side bearing.
                    if (firstGlyphOfLine)
                    {
                        offset.X = Math.Max(currentGlyph.LeftSideBearing, 0);
                        firstGlyphOfLine = false;
                    }
                    else
                        offset.X += Spacing + currentGlyph.LeftSideBearing;

                    // matrix calculations unrolled 
                    // 8 multiplications and up to 10 addition subtractions.

                    var m = offset;
                    m.X += currentGlyph.Cropping.X;
                    m.Y += currentGlyph.Cropping.Y;

                    if (flip)
                    {
                        if (flippedHorz)
                            m.X = msize.X - (m.X + currentGlyph.BoundsInTexture.Width);
                        if (flippedVert)
                            m.Y = msize.Y - (m.Y + currentGlyph.BoundsInTexture.Height);
                    }

                    var r = m - origin;
                    var p = r;

                    if (rotation != 0)
                    {
                        p.X = r.X * q.Y - r.Y * q.X;
                        p.Y = r.X * q.X + r.Y * q.Y;
                    }

                    dest = new Rectangle(
                        (int)(p.X * scale.X + position.X),
                        (int)(p.Y * scale.Y + position.Y),
                        (int)(currentGlyph.BoundsInTexture.Width * scale.X),
                        (int)(currentGlyph.BoundsInTexture.Height * scale.Y)
                        );

                    // this is extra, its a initial text cutting boundry box, the final exact version would go further up.
                    if (dest.Left > textboundryarea.Left && dest.Right < textboundryarea.Right && dest.Top > textboundryarea.Top && dest.Bottom < textboundryarea.Bottom)
                    {
                        // We call directly to Draw(,,) which will pass to draw internal
                        // so in this method i have bypassed drawinto, this is drawinto
                        // i could rewrite measure string and by pass that too because its inaccurate.
                        // however this belongs in monogame not in my own methods.
                        spriteBatch.Draw(tsf.Texture, dest, currentGlyph.BoundsInTexture, color, rotation, Vector2.Zero, effect, depth);
                    }

                    offset.X += currentGlyph.Width + currentGlyph.RightSideBearing;
                }
            }
        }
`

**Comment by tomspilman** [comment]
> no comments ?

Yeah...  the information you've written here is hard to digest, so i'm not surprised it was quiet. :)


> Ill begin by stating i have reduced the spritefont.DrawInto methods matrix calculations.

I think this is a perfect first PR to submit.  Just this will improve some performance when you do lots of individual text draw calls.  It will be easy to review and verify works.

Are there more simple fixes like this we can tackle first before a major re-shuffle of code?


> Spritefont class is a bit poorly structured in relation to Spritebatch

We are aware that there is a lot of code mixed between `SpriteFont`, `SpriteBatch`, and the internal `SpriteBatcher` classes.

My concern is making changes just for the sake of "organization" or making things "clean".  As long as the changes are specifically for performance and/or fixing bugs or incompatibilities with XNA then it is easer for us to accept.

**Comment by willmotil** [comment]
> Are there more simple fixes like this we can tackle first before a major re-shuffle of code?

I think so. 
But even these unrolled matrices are no were near ready to be implemented. This is only tested for functional equivalence not performance at all. I cant test that till its altered to fit the current develop build. 

Those recommendations are for consideration of possible problems due to the suggestion of adding the emjoi's as well as to show that this might be more complicated then anyone has really considered it could get im not really sure they are even possible to do without creating a new problem its just the only thing i could think of to suggest to decouple the two classes before they get derived from.
